// metaverse-api/lib/MessageBus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const EventDispatcher_1 = require("metaverse-rpc/lib/common/core/EventDispatcher");
/**
 * MessageBusClient provides an easy way to send and reveive peer to peer
 * messages thru the communications interface of Decentraland.
 *
 * The usage is simple, it implements a pub-sub pattern that shares messages
 * using the connection that already exists between peers.
 *
 * @example inside your ScriptableScene
 * ```ts
 *   let messageBusClient = await MessageBusClient.acquireEstateChannel(this)
 *   messageBusClient.on('a message name', (message) => {...})
 *   messageBusClient.emit('a message name', message)
 * ```
 */
class MessageBusClient extends EventDispatcher_1.EventDispatcher {
    constructor(api, id, busClientId) {
        super();
        this.api = api;
        this.id = id;
        this.busClientId = busClientId;
        api[`on${this.id}`]((message) => {
            if (this.busClientId !== message.sender) {
                super.emit(message.event, ...message.args);
            }
        });
    }
    static async acquireEstateChannel(system) {
        const busId = Math.random().toString(36);
        const { MessageBus } = await system.loadAPIs(['MessageBus']);
        const bus = await MessageBus.getEstateChannel();
        return new MessageBusClient(MessageBus, bus.id, busId);
    }
    emit(event, ...args) {
        this.api[this.id]({
            event,
            args,
            sender: this.busClientId
        });
        super.emit(event, ...args);
    }
}
exports.MessageBusClient = MessageBusClient;
